### PR TITLE
SkyeKiwi-Protocol Application & Amendment SkyePass deliverables

### PR DESCRIPTION
**Pull request by RoyTimes**
### Project Abstract

This amendment is made to change the deliverables on SkyePass https://github.com/w3f/Grants-Program/pull/212 

This amendment will make SkyePass dev process focus more on creating open-source re-usable technology for other developers in a similar shoe with us, and shift the focus away from the password manager itself. We want to make the core tech we have created under the hood the main product and the password manager as a cool tech demo. I think this will be more beneficial to the larger community and for us as well. Let us know if we would be able to do that or change the title/content of this application. 

### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( `project_name.md`) and updated.
- [X] The total funding amount of the project is below USD $30k for first-time grant applications and $100k for follow-up ones.
- [X] A BTC or Ethereum (DAI/USDT) address for the payment of the milestones is provided inside the application.
- [X] I have read and acknowledged the [terms and conditions](https://github.com/w3f/Grants-Program/blob/master/docs/T&Cs.md).
- [X] The software delivered for this grant will be released under an open-source license specified in the application.
- [X] The initial PR contains only one commit (squash and force-push if needed).
- [X] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery).

### CODE REVIEW

**Comment by RoyTimes** [comment]
@alxs @semuelle Please review the amended deliverables. It's quite some changes that is gonna fit more into our larger goal for the team. We came to realize that even though a decentralized password manger is cool but the technology that we have created under the hood to be more valuable beyond a toy. And at this time, we want to focus more on the technology and hire someone to work on the password manager after we have a well-established base tech layer. Let us know how you think and what's the best way to proceed. I would be happy to connect offline through a phone call if you'd like.

**Comment by semuelle** [comment]
Hi @RoyTimes, thanks for the submission. So, if I'm understanding correctly, this amendment will

- move the Desktop App deliverable to M2,
- add a `Core Encryption Library` to M1,
- remove the `UI/UX` and alpha invitations,
- move the start date to 2022, and
- move 1380 USD from M2 to M1?

Can you expand a bit on your comment about `focusing more on the technology`? What is it that you would like to achieve, in practical terms?

**Comment by alxs** [comment]
Hey, this is silly but could you fix the JSON at the beginning of the file? I think that's what's making the entire document so hard to read. Trying adding `]}` at the end.

File: applications/skyepass.md [hunk 1]
```diff
@@ -185,32 +185,30 @@ Besides private work for companies that cannot be shared, Song developed a simpl
...
-| 5.     | Testing                        | Unit Testing                                                 |
-### Milestone 2  — Ready for Beta Release
+| 0b.    | Documentations/Project Website | A minimalism project homepage with a user manual that documents how to use SkyePass from account creation to manage passwords, and install extensions and browser extensions. <br/>A developer documentation that describes the encryption mechanism in details; |
+| 1.     | Core Encryption Library        | Refined Core Encryption API to deliver three core APIs: `upstream`, `downstream` and `updateEncryptionSchema`. <br/>The encryption library should be able to: <br/>1. Post files onto an Infura IPFS pinning service<br/>2. Definitive length for metadata<br/>3. Able to interact with Substrate WASM smart contracts and a decentralized storage network.<br/>4. Reusable by other products other than SkyePass |
+| 2.     | Smart Contract                 | Deploy a smart contract for access control and file coordination. We will use Rust, ink! for smart contract development, along with tools like Redspot by [Patract Hub](https://patract.io/?lang=en-US) |
```
Comment (alxs) [inline applications/skyepass.md:203]: Why did you switch from implementing all 4 smart contracts under 'Ecosystem Fit' to just one?

**Comment by alxs** [comment]
@RoyTimes we could potentially discuss a parallel grant as long as this one's on hold, depending on what the core tech that you're building is. If you're interested send us a quick email to grants@web3.foundation with a brief project summary first.

**Comment by RoyTimes** [comment]
I have sent an email to grants@w3f to explain everything in more details. @semuelle Yes and more details sent via email. @alxs Yes. That would actually be the most ideal for us. Also detailed in the email: 

TL;DR is that we want to put SkyePass on hold until early next year. At the same time, for all the work we have been doing as part of the SkyePass should be transfer to the other application. Along with more work to be done by us. If the tech sounds good to you and you think will benefit the community, we would go ahead and consolidate the two application in two separate `.md` files and push to this PR.

File: applications/skyepass.md (hunk 1 above)
Comment (RoyTimes) [reply applications/skyepass.md:203]: Explained in the email. TL;DR version for that is we figured out another (a much better) way to deliver the identity management capacity needed and we are currently working with the on-chain treasury team on it. All things related to identity can be removed from the application now.

**Comment by RoyTimes** [comment]
Per the email, I have:
- create a new application for the SkyeKiwi Protocol and described works to be done within. 
- remove the deliverables related to the SkyeKiwi Protocol from the SkyePass application and indicate an est. start time of earlier next year. 
- Include all works has already been done by us. 

I don't want to write a 10-minute-read application for the SkyeKiwi Protocol and make it painful for you. So it's pretty concise. Let me know if you would like to learn more of any technical details of it. Always happy to answer those! The documentation site also includes more details: https://cdocs.skye.kiwi/#/

File: applications/skyepass.md (hunk 1 above)
Comment (RoyTimes) [reply applications/skyepass.md]: Also, have fixed the JSON thing in the most recent push. My editor was acting stupid :(

File: applications/skyepass.md [hunk 2]
```diff
@@ -184,46 +184,22 @@ Besides private work for companies that cannot be shared, Song developed a simpl
 | 3.     | Client Side UI/UX | an simple Electron UI/UX not wired up with logic yet         |
 
 
+### Milestone 1  — Desktop App & Browser Extension
```
Comment (alxs) [inline applications/skyepass.md:187]: Personally, I don't particularly mind this grant remaining open (at an appropriate price), though in the last discussion we had I think the consensus was rather that it should be closed, i.e. terminated in mutual agreement, if you're gonna work on something else. Doing so would also allow you more freedom when you've finished the other grant, and to be honest if your plans have changed in the last few months it's perfectly possible and fine if they do again.

File: applications/skyekiwi-protocol.md [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+* **Estimated duration:** 2 month
+* **FTE:**  2
+* **Costs:** 12,000USD
+
+| Number | Deliverable | Specification |
+| -----: | ----------- | ------------- |
+| 0a. | License | Apache 2.0 |
+| 0b. | Documentation | We will provide both **inline documentation** of the code and a basic **tutorial** that explains how a user can integrate the SkyeKiwi Protocol into their project. |
+| 0c. | Testing Guide | Core functions will be fully covered by unit tests to ensure functionality and robustness. In the guide, we will describe how to run these tests on both Node.js and Browser environment. |
+| 1. | Core Protocol | Implements the `upstream` `downstream` `updateEncryptionSchema` and `verifyProofOfAccess` APIs. |
```
Comment (alxs) [inline applications/skyekiwi-protocol.md:170]: This is JS?

File: applications/skyekiwi-protocol.md [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+* **Estimated duration:** 2 month
+* **FTE:**  2
+* **Costs:** 18,000USD
+
+| Number | Deliverable         | Specification                                                |
+| -----: | ------------------- | ------------------------------------------------------------ |
+|    0a. | License             | Apache 2.0                                                   |
+|    0b. | Documentation       | We will provide both **inline documentation** of the code and a basic **tutorial** that explains how a user can integrate the SkyeKiwi Protocol into their project. |
+|    0c. | Testing Guide       | Core functions will be fully covered by unit tests to ensure functionality and robustness. In the guide, we will describe how to run these tests on both Node.js, Browser and Rust environment. |
+|     1. | `pallet-secrets`    | Port the `ink!` smart contract to be a runtime pallet and integrate it to the SkyeKiwi Protocol source code. |
```
Comment (alxs) [inline applications/skyekiwi-protocol.md:188]: What's the purpose of implementing this twice, why don't you leave out the smart contract?

File: applications/skyekiwi-protocol.md [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+|     2. | Rust Implementation | Port the SkyeKiwi Protocol to Rust, so that it works between blockchain nodes as well. |
```
Comment (alxs) [inline applications/skyekiwi-protocol.md:189]: This would be deliverable 1 in M1? What do you mean by "between blockchain nodes"?

**Review by alxs — CHANGES_REQUESTED** [review]

File: applications/skyepass.md (hunk 2 above)
Comment (RoyTimes) [reply applications/skyepass.md:187]: Sure. We can come to agreement on this. We can replace the application for SkyePass to the SkyeKiwi Protocol. The reason that we are doing the replacement is actually a very rational decision: we realize that the secret transmitting technology underneath is more valuable than a password manager and for other project to reuse. For the next 12 month: we are building the protocol out and assemble a blockchain network. No switching now. Haha

File: applications/skyekiwi-protocol.md (hunk 5 above)
Comment (RoyTimes) [reply applications/skyekiwi-protocol.md:189]: No. M1 is implementing a client side version to run in browsers. M2-2 is a Rust implementation to run the protocol in Substrate. "Between blockchain nodes": encrypted by the identity key of nodes and able to share encrypted message between nodes with little coordination. However, the Rust version will be a straight & plain port, the way to encrypt with identity key and coordination will not be covered in this application.

File: applications/skyepass.md (hunk 2 above)
Comment (alxs) [reply applications/skyepass.md:187]: Whichever you prefer. I'm actually happy to accept this as is, let's see what the rest of the committee thinks. And in that case I agree with you ;)

File: applications/skyekiwi-protocol.md (hunk 4 above)
Comment (alxs) [reply applications/skyekiwi-protocol.md:188]: I was only worried this would mean you'll be doing some of the work twice, but those are very good reasons to intentionally do so.

**Review by alxs — APPROVED** [review]

**Review by Noc2 — APPROVED** [review]

**Review by semuelle — APPROVED** [review]
Like with the previous application, I still have doubts about the utility and target audience, but I am happy to be proven wrong.

**Comment by semuelle** [comment]
Hey @RoyTimes, how is the project going? Are you still on track for Milestone 2? Any unforeseen technical hurdles?

**Comment by RoyTimes** [comment]
Hi @semuelle! Time really flies. We are on track. For now, we have just finished a funding round and we are building the extended version of the protocol to support executable secrets and integrating a smart contract execution environment. It has been very hard to design but we managed to made it work. It was not included in the grant application but we have to do it to fit better to the larger scene. 

For milestone2, there are two major deliverables: 
1. We have finished the secret registry: https://github.com/skyekiwi/skyekiwi-network/tree/master/pallets/secrets
2. We have finished ~50% of the work for the enclave/Rust implementation of the protocol https://github.com/skyekiwi/skyekiwi-protocol-enclave - we kinda stuck at making the shamir secret sharing compatible with the client version but should be able to get back to it next week and building the protocol inside of Intel SGX environment is never easy. 

Generally, we are trying to make a delivery sometime mid Jan. Sorry for the delay. 

Also, on your previous doubts on the audience of the project - we are working on a lots of them: decentralized KYC and decentralized contract signing for starter.

**Comment by semuelle** [comment]
That's great to hear. Thanks for the update. Please keep us posted. And if you find you need to revise your timeline, you can always submit an amendment to the contract.

**Comment by alxs** [comment]
Hey @RoyTimes, any updates? What happened to https://github.com/skyekiwi/skyekiwi-protocol-enclave?

**Comment by RoyTimes** [comment]
@alxs Oh! I made it private as it has been merged with https://github.com/skyekiwi/skyekiwi-network Working 100hrs per week for the past month. This delivery is the next as long as we merge the latest PR. Things are on track with competing priorities of the VM and the substrate blockchain. The SkyeKiwi Protocol in enclave is the next part we are getting into.
